File: leetcode/problems/0006/main.py
class Solution:
    def convert(self, s: str, num_rows: int) -> str:

        mat = [[] for _ in range(num_rows)]
        delta = 1

        i = 0
        for c in s:
            mat[i].append(c)
            i = (i + delta) % num_rows
            delta = -1 * delta if i == 0 or i == num_rows - 1 else delta

        return "".join([c for l in mat for c in l])

File: leetcode/problems/0006/main.erl
-module(main).
-export([convert/2]).

append_elem([], _, _) -> [];
append_elem([Mat | Tail], 0, X) -> [lists:append(Mat, [X]) | Tail];
append_elem([Mat | Tail], I, X) -> [Mat | append_elem(Tail, I - 1, X)].

-spec convert(S :: unicode:unicode_binary(), NumRows :: integer()) -> unicode:unicode_binary().
convert(S, NumRows) ->
    Sl = binary_to_list(S),
    P = fun(C, {Mat, Delta, I}) -> 
        NewI = (I + Delta) rem NumRows,
        {
            append_elem(Mat, I, C),
            if
                (NewI == 0) or (NewI == NumRows - 1) -> Delta * (-1);
                true -> Delta 
            end,
            NewI
        }
    end,
    {Mat, _, _} = lists:foldl(P, {lists:duplicate(NumRows, []), 1, 0}, Sl),
    list_to_binary(lists:flatten(Mat)).
